src/codegen/compile.ml

```diff
@@ -7526,8 +7546,39 @@ and compile_prim_invocation (env : E.t) ae p es at =
       | Type.Promises -> assert false in
 
     let fun_sr, code1 = compile_exp env ae e1 in
+
+    (* we duplicate this pattern match to emulate pattern guards *)
+    let call_as_prim = match fun_sr, sort with
+      | SR.Const (_, Const.Fun (mk_fi, Const.PrimWrapper prim)), _ ->
+         begin match n_args, e2.it with
+         | 0, _ -> true
+         | 1, _ -> true
+         | n, PrimE (TupPrim, es) when List.length es = n -> true
+         | _, _ -> false
+         end
+      | _ -> false in
+
     begin match fun_sr, sort with
-      | SR.Const (_, Const.Fun mk_fi), _ ->
+      | SR.Const (_, Const.Fun (mk_fi, Const.PrimWrapper prim)), _ when call_as_prim ->
+         assert (sort = Type.Local);
+         (* Handle argument tuples *)
+         begin match n_args, e2.it with
+         | 0, _ ->
+           let sr, code2 = compile_prim_invocation env ae prim [] at in
+           sr,
+           code1 ^^
+           compile_exp_as env ae (StackRep.of_arity 0) e2 ^^
+           code2
+         | 1, _ ->
+           compile_prim_invocation env ae prim [e2] at
+         | n, PrimE (TupPrim, es) ->
+           assert (List.length es = n);
+           compile_prim_invocation env ae prim es at
+         | _, _ ->
+           (* ugly case; let's just call this as a function for now *)
```

Comment:
What would you do here. Evaluate the arguments once and project its components?

Reply:
Yes, something like that. Needs a bunch of locals. Overall, very rare I'd expect.

Reply:
> Note that the code falls though to a normal function call; it doesn't crash.

Do you mean a normal indirect or direct function call. I think because the next pattern matches on Const.Complicated (not `_`), it will actually fall through to the indirect call case. Is that what you intended?

Reply:
Ah, ok, I think its fine. (Just tested it myself).

Reply:
Actually, not it's not. In rare cases I think this will generate indirect calls that were previously direct, because of the bad fall-thru. See #3196 for a repro and suggested fix.